### PR TITLE
✨ Add cache on provider scope

### LINKED ISSUE #1647: Crontroller doesn't cache keystone tokens
/kind bug

**What steps did you take and what happened:**

We have a management cluster with the following resource count:
```console
$ kubectl get openstackclusters -A --no-headers | wc -l
4
$ kubectl get openstackmachines -A --no-headers | wc -l
67
```

When we update the `-v` flag to increase the verbosity, we can check the numbers of requests done to Keystone to get a new token.
```console
I0707 15:54:50.698548       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.701020       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.708509       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.712483       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.712982       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.715120       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.716857       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.739230       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.739574       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.753020       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.957990       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.960890       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.961781       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:50.964750       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:51.067514       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:51.672278       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:51.748524       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:51.954898       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:51.972360       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.168773       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.206942       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.345482       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.601283       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.810803       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.846344       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.886106       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:52.938195       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:53.026489       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens
I0707 15:54:54.846506       1 provider.go:127] OpenStack Request URL: POST [keystone_endpoint]/v3/auth/tokens  
```

We can see that we're doing many requests to keystone leading to some rate limiting when using public providers.

**What did you expect to happen:**

We can see that for each reconciliation on OpenStackClusters and OpenStackMachine we call `openstack.Authenticate` (see: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/pkg/scope/provider.go#L178). 
Maybe we can have some token caching mechanism ?

**Anything else you would like to add:**

I would be happy to discuss on finding the best solution to prevent this behavior and to implement a fix.

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): v0.7.1
- Cluster-API version: v1.3.5
- OpenStack version: stein
- Minikube/KIND version:
- Kubernetes version (use `kubectl version`): v1.25.6
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by alexandrevilain**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1647

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by mdbooth** [comment]
@alexandrevilain If you run `make lint-update` it normally fixes those things for you automatically.

**Comment by alexandrevilain** [comment]
I think it's ready for review.
Tested on my side, after the controller reboots, it makes somes API requests to keystone and then stop doing such request for 12 hours.

File: main.go [hunk 1]
```diff
@@ -156,6 +160,12 @@ func main() {
 		os.Exit(1)
 	}
 
+	if scopeCacheMaxSize <= 0 {
+		err := errors.New("--scope-cache-max-size should be > 0")
+		setupLog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
```
Comment (lentzi90) [inline main.go:167]: What do you think about allowing 0 cache size as a way to disable caching?
I'm not sure if that is something we need to support but it is often beneficial to be able to opt-out of new features

Comment (alexandrevilain) [reply main.go:167]: Good idea! 
Another benefit we can have is to debug without cache if something is going wrong

Comment (alexandrevilain) [reply main.go]: Updated, setting the flag to `--scope-cache-max-size=0` disable the cache.

**Review by lentzi90 — COMMENTED** [review]
Looks good!
/lgtm

**Review by mdbooth — COMMENTED** [review]
I'll try to get to this properly on Monday.

File: pkg/scope/scope.go [hunk 2]
```diff
@@ -20,14 +20,24 @@ import (
...
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha7"
 	"sigs.k8s.io/cluster-api-provider-openstack/pkg/clients"
 )
 
-// ScopeFactory is the default scope factory. It generates service clients which make OpenStack API calls against a running cloud.
-var ScopeFactory Factory = providerScopeFactory{}
+// NewScopeFactory creates the default scope factory. It generates service clients which make OpenStack API calls against a running cloud.
+func NewScopeFactory(maxCacheSize int) Factory {
```
Comment (seanschneeweiss) [inline pkg/scope/scope.go:32]: The package is already named `Scope`, so we could name this `NewFactory`.

```diff
- scope.NewScopeFactory
+ scope.NewFactory
```

File: pkg/scope/provider.go [hunk 3]
```diff
@@ -45,9 +48,11 @@ const (
 	caSecretKey     = "cacert"
 )
 
-type providerScopeFactory struct{}
+type providerScopeFactory struct {
+	cache *cache.LRUExpireCache
```
Comment (seanschneeweiss) [inline pkg/scope/provider.go:52]: We could make this cache more explicit, by e.g. naming it `clientCache`.
```suggestion
	clientCache *cache.LRUExpireCache
```

File: pkg/scope/provider.go [hunk 4]
```diff
@@ -106,6 +128,34 @@ func NewProviderScope(cloud clientconfig.Cloud, caCert []byte, logger logr.Logge
...
+		return nil, err
+	}
+
+	token, err := scope.ExtractToken()
+	if err != nil {
+		return nil, err
+	}
+
+	// compute the token expiration time
+	expiry := time.Until(token.ExpiresAt) / 2
```
Comment (seanschneeweiss) [inline pkg/scope/provider.go:153]: Out of curiosity:
Is using half expiry time good for time out of sync issues or what is the reason for choosing `/ 2` here?

File: pkg/scope/provider.go [hunk 5]
```diff
@@ -134,6 +184,14 @@ func (s *providerScope) NewLbClient() (clients.LbClient, error) {
 	return clients.NewLbClient(s.providerClient, s.providerClientOpts)
 }
 
+func (s *providerScope) ExtractToken() (*tokens.Token, error) {
+	client, err := openstack.NewIdentityV3(s.providerClient, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("can't create a new identity service client: %w", err)
```
Comment (seanschneeweiss) [inline pkg/scope/provider.go:190]:
```suggestion
		return nil, fmt.Errorf("create new identity service client: %w", err)
```

File: pkg/scope/provider.go [hunk 6]
```diff
@@ -106,6 +128,34 @@ func NewProviderScope(cloud clientconfig.Cloud, caCert []byte, logger logr.Logge
 	}, nil
 }
 
+func NewCachedProviderScope(cache *cache.LRUExpireCache, cloud clientconfig.Cloud, caCert []byte, logger logr.Logger) (Scope, error) {
+	key, err := getScopeCacheKey(cloud)
+	if err != nil {
+		return nil, fmt.Errorf("can't compute cloud config cache key: %w", err)
```
Comment (seanschneeweiss) [inline pkg/scope/provider.go:134]:
```suggestion
		return nil, fmt.Errorf("compute cloud config cache key: %w", err)
```

File: pkg/scope/provider.go [hunk 7]
```diff
@@ -63,10 +68,14 @@ func (providerScopeFactory) NewClientScopeFromMachine(ctx context.Context, ctrlC
 		caCert = defaultCACert
 	}
 
-	return NewProviderScope(cloud, caCert, logger)
+	if f.cache == nil {
```
Comment (seanschneeweiss) [inline pkg/scope/provider.go:71]: Should we check for `== 0` here, as it says "Setting this value to 0 means no cache."?

**Review by seanschneeweiss — COMMENTED** [review]
That is a great addition... We faced problems with Keystone as well, potentially related to too many auth requests, which is still being investigated at this point.

**Review by seanschneeweiss — COMMENTED** [review]
/lgtm

**Comment by lentzi90** [comment]
Any more comments @mdbooth ? Otherwise I think we can take this in :slightly_smiling_face:

**Review by EmilienM — APPROVED** [review]
/lgtm

**Pull request merged**